### ISSUE TITLE
FantasyPortrait is open-source now!

### ISSUE THREAD

**Issue by wangqiang9** [kijai/ComfyUI-WanVideoWrapper#1025]
The code, model and dataset of the expression control video generation [FantasyPortrait](https://github.com/Fantasy-AMAP/fantasy-portrait) based on Wan2.1 have been made public. 
Welcome to experience it!

@kijai Could you support merging into comfyui? Thank you.

**Comment by Rudra-ai-coder** [comment]
yeah this is a good addition @kijai

**Comment by kabachuha** [comment]
Do you plan to migrate it into Wan2.2?

**Comment by armynew** [comment]
> The code, model and dataset of the expression control video generation [FantasyPortrait](https://github.com/Fantasy-AMAP/fantasy-portrait) based on Wan2.1 have been made public. Welcome to experience it!
> 
> [@kijai](https://github.com/kijai) Could you support merging into comfyui? Thank you.

Is work on 8G vram or support 4 or 8 step lighting lora

**Comment by wangqiang9** [comment]
> Do you plan to migrate it into Wan2.2?

Sure, we are conducting relevant experiments. If we find that wan2.2 will significantly improve the quality of fantasyportrait than wan2.1, we will update it

**Comment by wangqiang9** [comment]
> > The code, model and dataset of the expression control video generation [FantasyPortrait](https://github.com/Fantasy-AMAP/fantasy-portrait) based on Wan2.1 have been made public. Welcome to experience it!
> > [@kijai](https://github.com/kijai) Could you support merging into comfyui? Thank you.
> 
> Is work on 8G vram or support 4 or 8 step lighting lora

After our current testing, fantasyportrait requires 20GB. If it is merged into comfyui, this will facilitate the use of WAN 2.1-related acceleration methods for fantasyportrait, which will significantly reduce the GPU memory used.

**Comment by Rudra-ai-coder** [comment]
> > Do you plan to migrate it into Wan2.2?
> 
> Sure, we are conducting relevant experiments. If we find that wan2.2 will significantly improve the quality of fantasyportrait than wan2.1, we will update it

yeah please do for wan2.2 as that has significant quality improvement

**Comment by HuntingSuccubus** [comment]
> The code, model and dataset of the expression control video generation [FantasyPortrait](https://github.com/Fantasy-AMAP/fantasy-portrait) based on Wan2.1 have been made public. Welcome to experience it!
> 
> [@kijai](https://github.com/kijai) Could you support merging into comfyui? Thank you.

how it compare to stableavatar? strange they have same release time

**Comment by wangqiang9** [comment]
> > The code, model and dataset of the expression control video generation [FantasyPortrait](https://github.com/Fantasy-AMAP/fantasy-portrait) based on Wan2.1 have been made public. Welcome to experience it!
> > [@kijai](https://github.com/kijai) Could you support merging into comfyui? Thank you.
> 
> how it compare to stableavatar? strange they have same release time

FantasyPortrait focuses on the direction of expression transfer from video to video and is capable of supporting multi-person scenes. However, stableavatar is an audio-driven video generation task. Our team previously open-sourced [FantasyTalking](https://fantasy-amap.github.io/fantasy-talking/), which is the first audio-driven video generation model based on Wan. Welcome to follow us.

**Comment by kijai** [comment]
Implemented in it's own branch for now due to the amount of dependencies in the face detection part mostly (and because I need to sleep), but works this much already:

https://github.com/user-attachments/assets/a1647996-9178-4f36-b37d-ff90f3620f2a

**Comment by Rudra-ai-coder** [comment]
thanks @kijai

**Comment by w2253543w** [comment]
谢谢@kijai

**Comment by kijai** [comment]
Merged to main now, the multitarget masking is still missing.

Tested some combinations and it works with MAGREF model as well as with Uni3C too:

https://github.com/user-attachments/assets/0b17ee2a-28e6-47f2-84f8-b0e74d0bc36a

**Comment by Rudra-ai-coder** [comment]
thanks the AGI @kijai  you are indeed AGI

**Comment by kijai** [comment]
Also with context windows for longer gens, 177 frames:

https://github.com/user-attachments/assets/02fd2e57-5cae-4b45-9fa3-fd1cb1047fd0

**Comment by caslix** [comment]
The repository has been updated and reinstalled, but it is still missing. I used both Latest and Nightly. What could be the problem? Thank you. The file requirements.txt has no changes

<img width="1620" height="564" alt="Image" src="https://github.com/user-attachments/assets/3dcd7a98-eff8-43c2-9845-8e5e8c0bbe25" />

**Comment by kijai** [comment]
> The repository has been updated and reinstalled, but it is still missing. I used both Latest and Nightly. What could be the problem? Thank you. The file requirements.txt has no changes
> 
> <img alt="Image" width="1620" height="564" src="https://private-user-images.githubusercontent.com/94461976/478043017-3dcd7a98-eff8-43c2-9845-8e5e8c0bbe25.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUxODE3ODcsIm5iZiI6MTc1NTE4MTQ4NywicGF0aCI6Ii85NDQ2MTk3Ni80NzgwNDMwMTctM2RjZDdhOTgtZWZmOC00M2MyLTk4NDUtOGU1ZThjMGJiZTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODE0VDE0MjQ0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0OWJmMGRkMGE3NzM3ZDNmOTZjZTZiNWYxNDRhMjFmMGZhMWQ1YjNiYzdmMDMzMjJmMzY2MDcxOTYxZTk1ZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.kaFpV2epA-etT4KvjrIUFW3BfsXLg8_1DJMg17-Cgyk">

If you're sure it's updated, you probably have some other fork or branch installed, make sure there's only one WanVideoWrapper -folder in your custom_nodes.

**Comment by caslix** [comment]
> 
> If you're sure it's updated, you probably have some other fork or branch installed, make sure there's only one WanVideoWrapper -folder in your custom_nodes.


I did install another fantasy_portrait branch, but I deleted it and reinstalled the main one, but there's some kind of trouble with the nodes.

**Comment by Rudra-ai-coder** [comment]
are the changes yet to be merged to main, after cloning the latest main still the following nodes are missing: 

<img width="1620" height="730" alt="Image" src="https://github.com/user-attachments/assets/0da4185b-9c4c-46a8-81d5-d36f26783634" />

**Comment by kijai** [comment]
> are the changes yet to be merged to main, after cloning the latest main still the following nodes are missing:
> 
> <img alt="Image" width="1620" height="730" src="https://private-user-images.githubusercontent.com/177262225/478079388-0da4185b-9c4c-46a8-81d5-d36f26783634.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUxODU1NDAsIm5iZiI6MTc1NTE4NTI0MCwicGF0aCI6Ii8xNzcyNjIyMjUvNDc4MDc5Mzg4LTBkYTQxODViLTljNGMtNDZhOC04MWQ1LWQzNmYyNjc4MzYzNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgxNFQxNTI3MjBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MmNhNDAxNmMwMmZhMDFkOTU4MmYzODUzOWU5NDE2NDc2NTFjZmE2NjY5MTdmNGJjMzIxZGQxNTJkMWI5MzllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jo5ZlGtK7LG4g5RoAS0_icwIXLWj0KTP3-C5c_e-y_c">

2 hours ago: https://github.com/kijai/ComfyUI-WanVideoWrapper/commit/24683f1dad761a70855869e3940bc3ed655223fb

**Comment by Rudra-ai-coder** [comment]
yeah let me try to clone the repo again, i can see the code in the main and in local clone but Comfy is not detecting the nodes not sure why

**Comment by 852wa** [comment]
fantasyportrait folder
model.py L:5
#from ..wanvideo.modules.attention import attention
nodes.py L:4
#from ..utils import log
model.py
Comment out from
#class SingleStreamBlockProcessor(nn.Module):
to
#return x
pip install opencv-python onnx onnxruntime-gpu scipy tqdm accelerate
After performing these processes, it worked.

**Comment by kijai** [comment]
> yeah let me try to clone the repo again, i can see the code in the main and in local clone but Comfy is not detecting the nodes not sure why

Actually I did wrap the import of the new nodes in try/except, if it fails to import the nodes wouldn't show up, maybe missing some dependency it has extra like mentioned above. No code changes needed though.

**Comment by Rudra-ai-coder** [comment]
@kijai can we add controlnet also? did u try that?

**Comment by Rudra-ai-coder** [comment]
> > yeah let me try to clone the repo again, i can see the code in the main and in local clone but Comfy is not detecting the nodes not sure why
> 
> Actually I did wrap the import of the new nodes in try/except, if it fails to import the nodes wouldn't show up, maybe missing some dependency it has extra like mentioned above. No code changes needed though.

this is fixed need to install onnx

**Comment by Rudra-ai-coder** [comment]
> [@kijai](https://github.com/kijai) can we add controlnet also? did u try that?

it works hehe what kinda of power is this

**Comment by kijai** [comment]
Check out what happens when you use both FantasyPortrait and MultiTalk:

https://github.com/user-attachments/assets/ef88058f-e962-49b7-9bfc-61ee4006a6e3

**Comment by Rudra-ai-coder** [comment]
> Check out what happens when you use both FantasyPortrait and MultiTalk:
> 
>  WanVideoFantasyPortrait_00001-audio.1.mp4

wooooow!!!!

**Comment by justforcivit** [comment]
> Check out what happens when you use both FantasyPortrait and MultiTalk:
> WanVideoFantasyPortrait_00001-audio.1.mp4

Amazing!  Could you post an example workflow showing how you combined FantasyPortrait and Multitalk?

**Comment by Rudra-ai-coder** [comment]
> > Check out what happens when you use both FantasyPortrait and MultiTalk:
> > WanVideoFantasyPortrait_00001-audio.1.mp4
> 
> Amazing! Could you post an example workflow showing how you combined FantasyPortrait and Multitalk?

just drag and drop the above video

**Comment by RedParisArts** [comment]
Any ideas why I'm getting this error when trying to use the Fantasy Portrait workflow?
^
ValueError("type fp8e4nv not supported in this architecture. The supported fp8 dtypes are ('fp8e4b15', 'fp8e5')")

On A6000

**Comment by Rudra-ai-coder** [comment]
Hi @kijai i was trying the above with fun control works great for the generation, but when used for the context window options after 81 frames the video becomes dark and noisier, this happening when using with Wan 2.1 fun control checkpoints  but works fine with regular I2V

**Comment by Xeverian** [comment]
Did the clean ComfyUI (0.3.50) + manager install today, then imported WanVideoWrapper. No matter what I do - it is missing the following nodes:

- LandmarksToImage
- FantasyPortraitModelLoader
- FantasyPortraitFaceDetector
- WanVideoAddFantasyPortrait

Any fix for this? Has anyone experienced the same lately? I've seen some comments above, but looks like the code has been chaned significantly since then

**Comment by kijai** [comment]
> Did the clean ComfyUI (0.3.50) + manager install today, then imported WanVideoWrapper. No matter what I do - it is missing the following nodes:
> 
> * LandmarksToImage
> * FantasyPortraitModelLoader
> * FantasyPortraitFaceDetector
> * WanVideoAddFantasyPortrait
> 
> Any fix for this? Has anyone experienced the same lately? I've seen some comments above, but looks like the code has been chaned significantly since then

They will fail to load if one of the new dependencies fails to load, for most people that has been `onnx` and `onnxruntime-gpu`:

`pip install -U onnx onnxruntime-gpu`

or for portable, in python_embeded folder:

`python.exe -m pip install -U onnx onnxruntime-gpu`

**Comment by shivansh-chanana** [comment]
> > Did the clean ComfyUI (0.3.50) + manager install today, then imported WanVideoWrapper. No matter what I do - it is missing the following nodes:
> > 
> > * LandmarksToImage
> > * FantasyPortraitModelLoader
> > * FantasyPortraitFaceDetector
> > * WanVideoAddFantasyPortrait
> > 
> > Any fix for this? Has anyone experienced the same lately? I've seen some comments above, but looks like the code has been chaned significantly since then
> 
> They will fail to load if one of the new dependencies fails to load, for most people that has been `onnx` and `onnxruntime-gpu`:
> 
> `pip install -U onnx onnxruntime-gpu`
> 
> or for portable, in python_embeded folder:
> 
> `python.exe -m pip install -U onnx onnxruntime-gpu`

I have these installed yet I am getting the same missing nodes error :(

**Comment by cochese9000** [comment]
> Check out what happens when you use both FantasyPortrait and MultiTalk:
> 
>  WanVideoFantasyPortrait_00001-audio.1.mp4

This all seems to have installed correctly, but when I run the workflow in this video, I get an execution error:

```
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "N:\airt\ComfyUI\execution.py", line 315, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "N:\airt\ComfyUI\execution.py", line 289, in _async_map_node_over_list
    await process_inputs(input_dict, i)
  File "N:\airt\ComfyUI\execution.py", line 277, in process_inputs
    result = f(**inputs)
             ^^^^^^^^^^^
  File "N:\airt\ComfyUI\custom_nodes\ComfyUI-WanVideoWrapper\fantasyportrait\nodes.py", line 123, in detect
    head_emo_feat_all, rect_list, landmark_list = get_emo_feature(numpy_list, face_aligner, pd_fpg_motion, device=device)
                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "N:\airt\ComfyUI\custom_nodes\ComfyUI-WanVideoWrapper\fantasyportrait\nodes.py", line 46, in get_emo_feature
    _, landmark_list, rect_list = det_landmarks(face_aligner, frame_list, comfy_pbar)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "N:\airt\ComfyUI\custom_nodes\ComfyUI-WanVideoWrapper\fantasyportrait\pd_fgc\pdf.py", line 175, in det_landmarks
    assert len(new_frame_list) > 0
           ^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

I updated comfyui and wanvideowrapper, but I still get it.  Anyone have any ideas?

**Comment by jakechai** [comment]
It means the face detector has failed. Try another video.

**Comment by cochese9000** [comment]
> It means the face detector has failed. Try another video.

Thankyou! That was the problem.

**Comment by Chargeuk** [comment]
> > > Check out what happens when you use both FantasyPortrait and MultiTalk:
> > > WanVideoFantasyPortrait_00001-audio.1.mp4
> > 
> > 
> > Amazing! Could you post an example workflow showing how you combined FantasyPortrait and Multitalk?
> 
> just drag and drop the above video

Hi @Rudra-ai-coder - I'm not sure why, but I am unable to do this? what browser and OS are you using? Thanks

**Comment by RuneGjerde** [comment]
> > just drag and drop the above video
> 
> Hi [@Rudra-ai-coder](https://github.com/Rudra-ai-coder) - I'm not sure why, but I am unable to do this? what browser and OS are you using? Thanks

Just download the video, and drag and drop it into ComfyUI as if it was a json workflow file (or image)
Should work fine (at least on Windows)
https://github.com/kijai/ComfyUI-WanVideoWrapper/issues/1025#issuecomment-3189224993

**Comment by Chargeuk** [comment]
> > > just drag and drop the above video
> > 
> > 
> > Hi [@Rudra-ai-coder](https://github.com/Rudra-ai-coder) - I'm not sure why, but I am unable to do this? what browser and OS are you using? Thanks
> 
> Just download the video, and drag and drop it into ComfyUI as if it was a json workflow file (or image) Should work fine (at least on Windows) [#1025 (comment)](https://github.com/kijai/ComfyUI-WanVideoWrapper/issues/1025#issuecomment-3189224993)

thanks @RunGjerde - I thought people were just dragging directly from the browser like you can with images. Interestingly the download and drag works on my Mac, but not on my windows box. No idea why, I must have set something up wrong in the install.
Thanks
